File: api/src/main/java/mb/pie/api/Output.java
package mb.pie.api;

import org.checkerframework.checker.nullness.qual.Nullable;

import java.io.Serializable;
import java.util.Objects;

public final class Output implements Serializable {
    public final @Nullable Serializable output;

    public Output(@Nullable Serializable output) {
        this.output = output;
    }

    @Override public boolean equals(Object o) {
        if(this == o) return true;
        if(o == null || getClass() != o.getClass()) return false;
        final Output that = (Output)o;
        return Objects.equals(this.output, that.output);
    }

    @Override public int hashCode() {
        return output != null ? output.hashCode() : 0;
    }

    @Override public String toString() {
        return "Output(output=" + this.output + ")";
    }
}

File: api/src/main/kotlin/mb/pie/api/Output.kt
package mb.pie.api

import java.io.Serializable

/**
 * Type for task outputs. Must be [Serializable], may be `null`.
 */
typealias Out = Serializable?

@Suppress("UNCHECKED_CAST", "NOTHING_TO_INLINE")
inline fun <O : Out> Out.cast() = this as O


/**
 * Wrapper for transient outputs; outputs that cannot be serialized. A transient output will be recreated when an attempt is made to
 * deserialize it, and then cached.
 */
interface OutTransient<out T : Any?> : Out {
  val v: T
  val consistent: Boolean
}

data class OutTransientImpl<out T : Any?>(
  @Transient override val v: T,
  @Transient override val consistent: Boolean = false
) : OutTransient<T> {
  constructor(v: T) : this(v, true)
}


/**
 * Specialization of [OutTransient], where a serializable value [e] is used for change detection through equality.
 */
interface OutTransientEquatable<out T : Any?, out E : Out> : OutTransient<T> {
  val e: E
}

data class OutTransientEquatableImpl<out T : Any?, out E : Out>(
  @Transient override val v: T,
  @Transient override val consistent: Boolean = false,
  override val e: E
) : OutTransientEquatable<T, E> {
  constructor(v: T, e: E) : this(v, true, e)
}
